build: Add vcpkg support

uristdwarf opened the pull request:
These commits add support for building vcpkg. This will now be used by default for production builds, instead of using system libraries (however they remain an option). This is to prevent compilation issues on newer compilers and old and/or buggy libraries being used.

It also makes changes to the CI, adding support for vcpkg, building on Ubuntu 22.04 and deployment to development apt repository for dev branch.

lgsilva3087 committed:
feat(build): Add presets for cmake with vcpkg

- Add presets for cmake with VCPKG.
- Add utility scripts for vcpkg setup and packages install/update.
- Set the packages to gather from vcpkg registry.
- Lock the dependency versions to be used for avoiding the
  consequences of undesired upgrades at unexpected moment.
- Add VCPKG usage to the documentation.
- Remove Hunter package manager
- Update gtests dependencies

Co-authored-by: guillex <guillex@leil.io>

Signed-off-by: Baldor <baldor@leil.io>

uristdwarf committed:
ci: Refactor CI
* Add Dockerfile from sfstests
* Seperate package installation script (for caching in Dockerfile)
* Use a fixed version of llvm.sh (don't wget it)
* Use vcpkg in Docker
* Disable caching CMake configuration
* Add disabled Ubuntu 22.04 tests

Signed-off-by: Urmas Rist <urmas@leil.io>

wg1k committed:
fix: Update github action for vcpkg

Signed-off-by: Baldor <baldor@leil.io>

dmga44 committed:
fix: Fix compilation in Windows

Signed-off-by: Dave <dave@leil.io>

wg1k committed:
fix(tests): Link to header version of spdlog and fmt

Also, specify the linking scope explicitly for other libs.

Signed-off-by: Baldor <baldor@leil.io>

uristdwarf committed:
ci: Fix run-unit-and-ganesha-tests with vcpkg and codespell

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf committed:
ci: Add deployment pipeline

This will use our internal deployment pipeline to build the packages and
push to developer repository, but ONLY on the dev branch.

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf committed:
build: Fix wrong path for examples

Due to where project("saunafs") is defined, it's adding the project name
to the folder with doc. This wasn't an issue in the past because project
name was defined later after the doc folder variable was declared.

This hardcodes the doc folder to be always
$CMAKE_INSTALL_FULL_DATAROOTDIR/doc/.

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf committed:
chore: Fix spelling

Signed-off-by: Urmas Rist <urmas@leil.io>

rolysr commented on tests/ci_build/run-build.sh:
```diff
@@ -25,12 +25,12 @@ usage() {
-	-DENABLE_DOCS=ON
 	-DENABLE_CLIENT_LIB=ON
-	-DENABLE_URAFT=ON
+	-DENABLE_DOCS=ON
 	-DENABLE_NFS_GANESHA=ON
-	-DGSH_CAN_HOST_LOCAL_FS=ON
 	-DENABLE_POLONAISE=OFF
+	-DENABLE_URAFT=ON
+	-DGSH_CAN_HOST_LOCAL_FS=ON
```
Why was it needed to change the order of cmake arguments here?

wg1k replied:
While not strictly needed, it is helpful to keep it sorted. It eases further comparisons (e.g., comparing the presets config).

rolysr commented on INSTALL.md:
```diff
@@ -3,9 +3,9 @@
-- Linux (x86_64)
-  - Ubuntu 24.04 (Noble)
-  - Ubuntu 22.04 (Jammy)
+-   Linux (x86_64)
+    -   Ubuntu 24.04 (Noble)
+    -   Ubuntu 22.04 (Jammy)
```
Why adding this extra spaces between the `-`? Was this for better readability on some display? Or some  expected indentation?

wg1k replied:
This one is the IDE following the `.editorconfig` of the project's root for .md files.
After indentation is easier to follow the items when the dash is below the first letter of the owning item, so the formatter introduced those spaces for complying both with the 4 spaces spacing rule (from editor config) and this readability nuance. It only affects the raw text source code, as both versions will be rendered same way in HTML.
IMO it is OK to allow this change. Do you agree?

rolysr replied:
Good

rolysr commented on INSTALL.md:
````diff
@@ -16,62 +16,62 @@ system.
-       --keyserver hkps://keyserver.ubuntu.com \
-       --receive-keys 0xA80B96E2C79457D4
-   ```
+    ```shell
+    gpg --no-default-keyring \
+        --keyring /usr/share/keyrings/saunafs-archive-keyring.gpg \
+        --keyserver hkps://keyserver.ubuntu.com \
+        --receive-keys 0xA80B96E2C79457D4
+    ```
````
I noticed that this extra spaces were added in some parts of this `INSTALL.md` file. Are all this indentation changes expected, even when the content is not being changed?

wg1k replied:
This one is the IDE following the `.editorconfig` of the project's root for `.md` files.

lgsilva3087 commented on CMakeLists.txt:
```diff
@@ -3,17 +3,8 @@ cmake_minimum_required(VERSION 3.20)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
-# Allows to find libraries not installed by Hunter (like libcrcutil)
 list(APPEND CMAKE_PREFIX_PATH "/usr/local" "/usr")
```
This line was added for Hunter, please check if it is still needed.

wg1k replied:
done

lgsilva3087 commented on cmake/Libraries.cmake:
```diff
@@ -76,12 +69,10 @@ else()
-
-set(BOOST_MIN_VERSION "1.48.0")
-find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS filesystem iostreams program_options system)
+#set(Boost_NO_SYSTEM_PATHS ON)
```
Consider removing the comment if it is not needed.

wg1k replied:
done

lgsilva3087 commented on cmake/SharedLibraries.cmake:
```diff
@@ -52,8 +52,8 @@ function(shared_target_link_libraries TARGET)
 	endforeach()
 
-	target_link_libraries(${TARGET} ${libraries_static})
+	target_link_libraries(${TARGET} PRIVATE ${libraries_static})
```
Was the change to PRIVATE needed?

wg1k replied:
actually not, reverting

lgsilva3087 commented on src/chunkserver/CMakeLists.txt:
```diff
@@ -3,25 +3,25 @@ add_definitions(-DSAUNAFS_MAX_FILES=10000)
 
 # Boost libraries used for plugins
-find_package(Boost REQUIRED COMPONENTS filesystem system)
-include_directories(${Boost_INCLUDE_DIRS})
+#find_package(Boost REQUIRED COMPONENTS filesystem system)
+#include_directories(${Boost_INCLUDE_DIRS})
```
Consider removing these comments if not needed.

wg1k replied:
done

lgsilva3087 commented on src/chunkserver/CMakeLists.txt:
```diff
@@ -3,25 +3,25 @@ add_definitions(-DSAUNAFS_MAX_FILES=10000)
-target_link_libraries(chunkserver chunkserver-common safsprotocol sfscommon
-    ${Boost_LIBRARIES}
+target_link_libraries(chunkserver PRIVATE chunkserver-common safsprotocol sfscommon
+#    ${Boost_LIBRARIES}
```
Consider remove.

wg1k replied:
done

lgsilva3087 commented on src/chunkserver/chunkserver-common/CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR})
 shared_add_library(chunkserver-common ${CHUNKSERVER_PLUGINS_SOURCES})
-target_link_libraries(chunkserver-common safsprotocol sfscommon
-    ${Boost_LIBRARIES} ${ADDITIONAL_LIBS})
+target_link_libraries(chunkserver-common PRIVATE safsprotocol sfscommon
```
${Boost_LIBRARIES} could be not defined (empty) here, because we are using the CONFIG approach in the call to find_package. Please double check.

wg1k replied:
Boost_LIBRARIES: Boost::filesystem;Boost::iostreams;Boost::program_options;Boost::system

lgsilva3087 commented on src/metadump/CMakeLists.txt:
```diff
@@ -2,5 +2,5 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR})
 aux_source_directory(. METADUMP_SOURCES)
 add_executable(sfsmetadump ${METADUMP_SOURCES})
-target_link_libraries(sfsmetadump)
+target_link_libraries(sfsmetadump PRIVATE)
```
This line looks unneeded. Please check.

wg1k replied:
reverted

lgsilva3087 commented on src/slogger/CMakeLists.txt:
```diff
@@ -1,4 +1,9 @@
+if(MINGW)
+    target_link_libraries(slogger PRIVATE fmt::fmt-header-only)
+else()
+    target_link_libraries(slogger PRIVATE  fmt::fmt-header-only spdlog::spdlog_header_only)
```
This works, but could make the building slower than using the static libs.

wg1k replied:
Addressed here: 8040f5f7281408e30a1669e79faa36c3b444a2d2

uristdwarf replied:
This caused issues with spdlog printing to stdout, please see 224def354eed0935cf5b88adbcad23ea1eb81476.

lgsilva3087 commented on src/unittests/CMakeLists.txt:
```diff
@@ -15,9 +13,10 @@ endif()
+  target_link_libraries(unittest_${unittest_name} PRIVATE
         ${ENABLE_WHOLE_ARCHIVE} ${unittest_name}_unittest ${DISABLE_WHOLE_ARCHIVE} ${${unittest_name}_UNITTEST_LINKLIST}
-        unittests-helpers unittest-mocks ${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY})
+        unittests-helpers unittest-mocks ${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} GTest::gtest GTest::gtest_main pthread)
```
${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} could be empty here (and not needed). Please double check.

wg1k replied:
These are not empty :)

lgsilva3087 replied:
Good.

lgsilva3087 commented on src/uraft/CMakeLists.txt:
```diff
@@ -7,7 +7,7 @@ set(URAFT_SOURCES ${URAFT_SOURCES} ${CMAKE_SOURCE_DIR}/src/common/time_utils.cc)
 add_executable(saunafs-uraft ${URAFT_MAIN} ${URAFT_SOURCES})
 
-target_link_libraries(saunafs-uraft ${Boost_LIBRARIES} ${RT_LIBRARY} pthread)
+target_link_libraries(saunafs-uraft PRIVATE ${Boost_LIBRARIES} ${RT_LIBRARY} pthread)
```
Please check the content of ${Boost_LIBRARIES} here.

wg1k replied:
`Boost::filesystem;Boost::iostreams;Boost::program_options;Boost::system`

lgsilva3087 commented on tests/CMakeLists.txt:
```diff
@@ -36,5 +34,6 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR}) # test_cases.h is located here
 add_dependencies(saunafs-tests list_of_test_cases)
-target_link_libraries(saunafs-tests ${GTEST_BOTH_LIBRARIES} ${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY})
+target_link_libraries(saunafs-tests PRIVATE GTest::gtest GTest::gtest_main)
+target_link_libraries(saunafs-tests PRIVATE ${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY})
```
The same here. My concern is the change of now using CONFIG in the find_package calls.

wg1k replied:
Boost_SYSTEM_LIBRARY: Boost::system
Boost_FILESYSTEM_LIBRARY: Boost::filesystem

lgsilva3087 reviewed (commented):
Please see my comments.

wg1k committed:
revert(build): Bring back link scope to defaults

The 0ee0ef67 commit introduced changes that are out of the scope of
this PR.

Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
fix(build): Apply PR suggestions

Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
fix(build): Add missing header

The build broke and could be run with the header-only version of spdlog,
but that was a fleeting patch. This commit fixes the cause of the
failure and reverts the patch.

Signed-off-by: Baldor <baldor@leil.io>

uristdwarf committed:
fix: Fix metarestore polluting stderr

Because of the changes to static linking of spdlog from vcpkg, it is
automatically adding a default logger, that logs to stdout. This is not
affecting main, because we removed all logs at setup. However, this was
not the case for metarestore, which means stdout was polluted with logs,
causing some tests to fail to properly parse files.

This change fixes that by dropping all logs on metarestore, before
adding stderr and syslog logs.

Signed-off-by: Urmas Rist <urmas@leil.io>

lgsilva3087 reviewed (approved):
LGTM
